src/core/toga/widgets/canvas.py

```diff
@@ -11,99 +11,77 @@ class Canvas(Widget):
         id (str):  An identifier for this widget.
         style (:obj:`Style`): An optional style object. If no
             style is provided then a new one will be created for the widget.
-        on_draw (``callable``): Function to draw on the canvas.
         factory (:obj:`module`): A python module that is capable to return a
             implementation of this class with the same name. (optional &
             normally not needed)
     """
 
-    def __init__(self, id=None, style=None, on_draw=None, factory=None):
+    def __init__(self, id=None, style=None, factory=None):
         super().__init__(id=id, style=style, factory=factory)
 
         # Create a platform specific implementation of Canvas
         self._impl = self.factory.Canvas(interface=self)
-        self.on_draw = on_draw
-
-    def set_context(self, context):
-        """The context of the Canvas to pass through from a callable function
-
-        Args:
-            context (:obj:'context'): The context to pass
-
-        """
-        self._impl.set_context(context)
-
-    @property
-    def on_draw(self):
-        """The callable function to draw on a Canvas
-
-        Creates a context or saves the current context, draws the operations
-        using the passed function, then renders the drawing, and finally
-        restores the saved context. The top left corner of the canvas must be
-        painted at the origin of the context and is sized using the rehint()
-        method.
-
-        """
-        return self._on_draw
-
-    @on_draw.setter
-    def on_draw(self, handler):
-        self._on_draw = handler
-
-        if self._on_draw:
-            self._impl.set_on_draw(self._on_draw)
-
-    # Line Styles
-
-    def line_width(self, width=2.0):
-        """Set width of lines
 
+        self.context_root = []
+        self._impl.set_context_root(self.context_root)
+        self.drawing_objects = self.context_root
```

Comment:
What's the difference between the "context root" and the drawing objects?

Reply:
self.drawing_objects gets switched to the current stack of drawing objects. If you aren't using a context manager then that is the context_root. If you are, then it becomes the drawing_objects stack that is part of that Context.